File: app/routers/databases.py
"""Database management router (refactored).

Handles database creation, switching, import/export.
Uses domain types where applicable.
"""
from fastapi import APIRouter, HTTPException, Depends, File, Form, UploadFile
from fastapi.responses import FileResponse
from pathlib import Path
import tempfile
import shutil

from ..models import DatabaseCreate, User
from ..domain import DatabaseNotFoundError, DuplicateDatabaseError
from .auth import get_current_user

# Import legacy db - database management is infrastructure-level
from .. import database as db

router = APIRouter(prefix="/api/databases", tags=["Databases"])


def _db_error_to_http(error: Exception) -> HTTPException:
    """Convert domain errors to HTTP exceptions."""
    if isinstance(error, DatabaseNotFoundError):
        return HTTPException(status_code=404, detail=error.message)
    elif isinstance(error, DuplicateDatabaseError):
        return HTTPException(status_code=409, detail=error.message)
    else:
        return HTTPException(status_code=500, detail=str(error))


@router.get("")
async def list_databases(user: User = Depends(get_current_user)):
    """List all available databases for current user."""
    databases = await db.list_databases(user.id)
    active = db.get_active_db_name(user.id)
    return {"databases": databases, "active": active}


@router.get("/check-name/{name}")
async def check_database_name(name: str, user: User = Depends(get_current_user)):
    """Check if a database name is available."""
    databases = await db.list_databases(user.id)
    exists = any(d['name'] == name for d in databases)
    return {"available": not exists, "name": name}


@router.post("")
async def create_database(data: DatabaseCreate, user: User = Depends(get_current_user)):
    """Create a new database."""
    try:
        database = await db.create_database(user.id, data.name)
        return database
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{name}/switch")
async def switch_database(name: str, user: User = Depends(get_current_user)):
    """Switch to a different database."""
    success = await db.switch_database(user.id, name)
    if not success:
        raise HTTPException(status_code=404, detail=f"Database '{name}' not found")
    return {"status": "ok", "active": name}


@router.put("/{name}/rename")
async def rename_database(name: str, data: DatabaseCreate, user: User = Depends(get_current_user)):
    """Rename a database."""
    try:
        database = await db.rename_database(user.id, name, data.name)
        return database
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.delete("/{name}")
async def delete_database(name: str, user: User = Depends(get_current_user)):
    """Delete a database."""
    try:
        success = await db.delete_database(user.id, name)
        if not success:
            raise HTTPException(status_code=404, detail=f"Database '{name}' not found")
        return {"status": "ok"}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{name}/export")
async def export_database(name: str, user: User = Depends(get_current_user)):
    """Export a database file for download."""
    try:
        export_path = await db.export_database(user.id, name)
        return FileResponse(
            export_path,
            filename=export_path.name,
            media_type="application/octet-stream"
        )
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/import")
async def import_database(
    file: UploadFile = File(...),
    name: str = Form(...),
    user: User = Depends(get_current_user)
):
    """Import a database file."""
    # Save uploaded file temporarily
    with tempfile.NamedTemporaryFile(delete=False, suffix=".db") as tmp:
        shutil.copyfileobj(file.file, tmp)
        tmp_path = Path(tmp.name)
    
    try:
        database = await db.import_database(user.id, tmp_path, name)
        return database
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    finally:
        tmp_path.unlink(missing_ok=True)

File: frontend/src/api/databases.ts
/**
 * Database API functions
 */
import api from './client';

export interface DatabaseInfo {
  uuid: string;
  name: string;
  created_at?: string;
  updated_at?: string;
  node_count?: number;
  page_count?: number;
  asset_count?: number;
  size_bytes?: number;
  is_active?: boolean;
  is_shared?: boolean;
}

export interface DatabaseListResponse {
  databases: DatabaseInfo[];
  active: string | null;
}

export interface DatabaseCheckResponse {
  available: boolean;
  name: string;
}

/**
 * List all databases for the current user
 */
export async function listDatabases(): Promise<DatabaseListResponse> {
  const response = await api.get('/databases');
  return response.data;
}

/**
 * Check if a database name is available
 */
export async function checkDatabaseName(name: string): Promise<DatabaseCheckResponse> {
  const response = await api.get(`/databases/check-name/${encodeURIComponent(name)}`);
  return response.data;
}

/**
 * Create a new database
 */
export async function createDatabase(name: string): Promise<DatabaseInfo> {
  const response = await api.post('/databases', { name });
  return response.data;
}

/**
 * Switch to a different database by workspace UUID
 */
export async function switchDatabase(uuid: string): Promise<{ status: string; active: string }> {
  const response = await api.post(`/databases/${encodeURIComponent(uuid)}/switch`);
  return response.data;
}

/**
 * Delete a database
 */
export async function deleteDatabase(name: string): Promise<{ status: string }> {
  const response = await api.delete(`/databases/${encodeURIComponent(name)}`);
  return response.data;
}

/**
 * Rename a database
 */
export async function renameDatabase(oldName: string, newName: string): Promise<DatabaseInfo> {
  const response = await api.put(`/databases/${encodeURIComponent(oldName)}/rename`, { name: newName });
  return response.data;
}

/**
 * Import a database from file
 */
export async function importDatabase(name: string, file: File): Promise<DatabaseInfo> {
  const formData = new FormData();
  formData.append('name', name);
  formData.append('file', file);
  const response = await api.post('/databases/import', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return response.data;
}

/**
 * Get export URL for a database
 */
export function getDatabaseExportUrl(name: string): string {
  return `/api/databases/${encodeURIComponent(name)}/export`;
}

/**
 * Get all settings for the current database
 */
export async function getSettings(): Promise<Record<string, string>> {
  const response = await api.get('/settings');
  return response.data;
}

/**
 * Get a specific setting
 */
export async function getSetting(key: string): Promise<string | null> {
  const settings = await getSettings();
  return settings[key] ?? null;
}

/**
 * Set a setting value
 */
export async function setSetting(key: string, value: unknown): Promise<void> {
  await api.put(`/settings/${encodeURIComponent(key)}`, { value });
}
